src/onnx/parse_attention.cpp

```diff
@@ -0,0 +1,967 @@
+                                Use (batch, 1, max_sequence_length, max_sequence_length) for shapes of size 4");
+            }
+            inferred_out.max_sequence_length = mask_index_lens.at(2);
+            inferred_out.index_pad           = mask_pad::raw;
+            MIGRAPHX_THROW("Attention: Mask_index 4D Megatron masking not supported");
+        }
+        else
+        {
+            MIGRAPHX_THROW(
+                "Attention: Mask_index Require shape of size either 1, 2, 3, 4 dimensions");
+        }
+    }
+
+    static void handle_mask_index(const std::vector<instruction_ref>& args,
+                                  struct attention_attr& /*attr_out*/,
+                                  struct attention_inferred& inferred_out,
+                                  std::vector<instruction_ref>& output_arg_vec)
+    {
+        instruction_ref mask_index;
+        if(check_and_return_arg(args, 3, mask_index))
+        {
+            auto mask_index_shape       = mask_index->get_shape();
+            const auto& mask_index_lens = mask_index_shape.lens();
+
+            if(mask_index_shape.type() != migraphx::shape::int32_type)
+            {
+                MIGRAPHX_THROW("Attention: Mask_Index type must be int32 type");
+            }
+            check_mask_index_shapes(mask_index_lens, inferred_out);
+            inferred_out.has_attn_mask = true;
+            output_arg_vec.push_back(mask_index);
+        }
+    }
+
+    static void handle_past(const std::vector<instruction_ref>& args,
+                            const struct attention_attr& attr_out,
+                            struct attention_inferred& inferred_out,
+                            std::vector<instruction_ref>& output_arg_vec)
+    {
+        instruction_ref past;
+        if(check_and_return_arg(args, 4, past))
+        {
+            inferred_out.has_past_input = true;
+            if(args.size() != 7)
+            {
+                MIGRAPHX_THROW("Attention: Past input requires past_sequence_length to be set");
+            }
+
+            const auto past_shape = past->get_shape();
+            const auto& past_lens = past_shape.lens();
+            if((past_lens.at(0) != inferred_out.batch_size) or
+               (past_lens.at(1) != attr_out.num_heads) or
+               (past_lens.at(3) != inferred_out.query_size) or (past_lens.size() != 4))
+            {
+                MIGRAPHX_THROW(
+                    "Attention: Past shape must be (batch, num_heads, max_sequence_length, "
+                    "head_size)\n OR when past_present_share_buffer set (batch, num_heads, "
+                    "total_sequence, head_size)");
+            }
+
+            if(attr_out.past_present_share_buffer)
+            {
+                inferred_out.total_sequence_length = past_lens.at(2);
+            }
+            else
+            {
+                // TODO: Handle case where non shared buffer in past sequence length
```

Comment:
You should `throw` here and remove the rest of the code below.

Reply:
Done. I'll these checks back in once I add the full end to end for these other inputs/attributes in other PRs when needed